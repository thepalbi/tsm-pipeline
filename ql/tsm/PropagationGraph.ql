/**
 * A query for computing triples in propagation graphs.
 */

import PropagationGraphs

class AllPackagesAreInteresting extends InterestingPackageForSinks, InterestingPackageForSources {
  AllPackagesAreInteresting() { exists(API::moduleImport(this)) }
}

from DataFlow::Node src, DataFlow::Node san, DataFlow::Node snk
where triple(src, san, snk)
select rep(src, false), rep(san, false), rep(snk, true)
