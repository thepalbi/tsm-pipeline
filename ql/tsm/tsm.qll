import javascript

import NodeRepresentation
import tsm_repr_pred

module TSM {
    private import TsmRepr

    predicate isSourceCandidate(DataFlow::Node nd) {
          nd instanceof DataFlow::CallNode or
          nd instanceof DataFlow::PropRead or
          nd instanceof DataFlow::ParameterNode
      }
  
      predicate isSanitizerCandidate(DataFlow::Node nd) {
        nd instanceof DataFlow::CallNode
      }
  
      predicate isSinkCandidate(DataFlow::Node nd) {
        (
          exists(DataFlow::InvokeNode invk |
            nd = invk.getAnArgument()
            or
            nd = invk.(DataFlow::MethodCallNode).getReceiver()
          )
          or
          nd = any(DataFlow::PropWrite pw).getRhs()
        )
      }

    string rep(DataFlow::Node node){
        result = candidateRep(node,_, _)
    }

    predicate isSink(DataFlow::Node node, float score){
        isSinkCandidate(node) and
        (exists(rep(node)) and   score = sum(doGetReprScore(rep(node), "snk"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    predicate isSource(DataFlow::Node node, float score){
        isSourceCandidate(node) and
        (exists(rep(node)) and   score = sum(doGetReprScore(rep(node), "src"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    predicate isSanitizer(DataFlow::Node node, float score){
        isSanitizerCandidate(node) and
        (exists(rep(node)) and
        score = sum(doGetReprScore(rep(node), "san"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    float doGetReprScore(string repr, string t){
        result = TsmRepr::getReprScore(repr, t)
   }    
}