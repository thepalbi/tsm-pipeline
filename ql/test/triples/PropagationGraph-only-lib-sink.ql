/**
 * A query for computing triples in propagation graphs.
 */

import tsm.PropagationGraphs

class AllPackagesAreInteresting extends InterestingPackageForSources {
  AllPackagesAreInteresting() { exists(API::moduleImport(this)) }
}

class LibIsInteresting extends InterestingPackageForSinks {
  LibIsInteresting() { this in ["lib"] }
}

from DataFlow::Node src, DataFlow::Node san, DataFlow::Node snk
where triple(src, san, snk)
select rep(src, false), rep(san, false), rep(snk, true)
