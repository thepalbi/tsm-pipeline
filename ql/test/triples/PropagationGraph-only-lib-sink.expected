| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (member * (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member * (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member * (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member * (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm (parameter 0 (member callback *)))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (member f (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f (parameter parm *)) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (member f *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (member f *) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (member f *) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback (root https://www.npmjs.com/package/lib)))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback *))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm (parameter 0 (member callback *))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (parameter parm *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (parameter parm *) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (parameter parm *) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 *))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop *) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink2 *)) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member * (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (member prop *) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink2 *)) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source (root https://www.npmjs.com/package/lib))) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member * (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member * (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (member prop *) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink2 *)) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize (root https://www.npmjs.com/package/lib))) | (parameter 0 (member sink *)) |
| (return (member source *)) | (return (member sanitize *)) | (member * (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize *)) | (member * (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member sanitize *)) | (member prop (parameter 0 (member * (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib)))) |
| (return (member source *)) | (return (member sanitize *)) | (member prop (parameter 0 (member sink2 *))) |
| (return (member source *)) | (return (member sanitize *)) | (member prop *) |
| (return (member source *)) | (return (member sanitize *)) | (parameter 0 (member * (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize *)) | (parameter 0 (member sink2 (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize *)) | (parameter 0 (member sink2 *)) |
| (return (member source *)) | (return (member sanitize *)) | (parameter 0 (member sink (root https://www.npmjs.com/package/lib))) |
| (return (member source *)) | (return (member sanitize *)) | (parameter 0 (member sink *)) |
